python/tank/bootstrap/manager.py

```diff
@@ -921,39 +922,78 @@ def _start_engine(self, tk, engine_name, entity, progress_callback=None):
 
         # perform absolute import to ensure we get the new swapped core.
         import tank
+        is_shotgun_engine = engine_name == constants.SHOTGUN_ENGINE_NAME
 
         # handle legacy cases
-        if engine_name == constants.SHOTGUN_ENGINE_NAME and is_version_older(tk.version, "v0.18.77"):
-
-            # bootstrapping into a shotgun engine with an older core
-            # we perform this special check to make sure that we correctly pick up
-            # the shotgun_xxx.yml environment files, even for older cores.
-            # new cores handles all this inside the tank.platform.start_shotgun_engine
-            # business logic.
-            log.debug(
-                "Target core version is %s. Starting shotgun engine via legacy pathway." % tk.version
-            )
-
-            if entity is None:
-                raise TankBootstrapError(
-                    "Legacy shotgun environments do not support bootstrapping into a site context."
-                )
-
-            # start engine via legacy pathway
-            # note the local import due to core swapping.
-            from tank.platform import engine
-            engine = engine.start_shotgun_engine(tk, entity["type"], ctx)
-
+        if is_shotgun_engine and is_version_older(tk.version, "v0.18.77"):
+            engine = self._legacy_start_shotgun_engine(tk, engine_name, entity, ctx)
         else:
             # no legacy cases
-            engine = tank.platform.start_engine(engine_name, tk, ctx)
+            try:
+                engine = tank.platform.start_engine(engine_name, tk, ctx)
+            except TankEngineInitError as exc:
+                # It's possible that a tk-core is being used that didn't come from
+                # the app_store. This might be the case where a site config has been
+                # locked off, and populated with a tk-core cloned from Github. In that
+                # case, it'll have "HEAD" as its version. In that situation, we don't
+                # actually know if this is a "new" core, or something super old. Given
+                # that, we need to try going the legacy route here just to see if we
+                # might actually have an old core.
+                if is_version_head(tk.version) and is_shotgun_engine:
+                    try:
+                        engine = self._legacy_start_shotgun_engine(tk, engine_name, entity, ctx)
+                    except Exception:
+                        # We'll want to raise the original exception here. We don't
+                        # really know whether this is a legacy core or not, so if both
+                        # the legacy and new code paths failed, we'll raise the exception
+                        # from the new code path.
+                        log.debug(
```

Comment:
Is it worth making this more than debug so it is easier to find in the logs if it comes up?